optaplanner-wb-screens/optaplanner-wb-guided-rule-editor/optaplanner-wb-guided-rule-editor-client/src/main/java/org/optaplanner/workbench/screens/guidedrule/client/plugin/ActionPluginUtils.java

```diff
@@ -0,0 +1,84 @@
+
+import java.util.ArrayList;
+import java.util.Collection;
+import javax.enterprise.context.ApplicationScoped;
+import javax.inject.Inject;
+
+import org.jboss.errai.common.client.api.Caller;
+import org.jboss.errai.common.client.api.RemoteCallback;
+import org.optaplanner.workbench.screens.guidedrule.client.widget.ScoreHolderGlobalAware;
+import org.optaplanner.workbench.screens.guidedrule.model.AbstractActionMultiConstraintBendableMatch;
+import org.optaplanner.workbench.screens.guidedrule.model.ActionBendableHardConstraintMatch;
+import org.optaplanner.workbench.screens.guidedrule.model.ActionBendableSoftConstraintMatch;
+import org.optaplanner.workbench.screens.guidedrule.model.BendableScoreLevelsWrapper;
+import org.optaplanner.workbench.screens.guidedrule.service.ScoreHolderService;
+import org.uberfire.backend.vfs.Path;
+import org.uberfire.mvp.Command;
+
+@ApplicationScoped
+public class ActionPluginUtils {
+
+    @Inject
+    private Caller<ScoreHolderService> scoreHolderService;
+
+    public void initScoreHolderAwarePlugin( final Path projectPath,
+                                            final ScoreHolderGlobalAware scoreHolderGlobalAware,
+                                            final Collection supportedScoreHolderTypes ) {
+        scoreHolderService.call( new RemoteCallback<Collection<String>>() {
+            @Override
+            public void callback( final Collection<String> scoreHolderFqns ) {
+                if ( scoreHolderFqns.isEmpty() ) {
+                    scoreHolderGlobalAware.scoreHolderGlobalIssueDetected( "'scoreHolder' global not defined. The global variable is created automatically when a Planning Solution is created. Make sure to create a planning solution before proceeding to rule definition." );
+                } else if ( scoreHolderFqns.size() > 1 ) {
+                    scoreHolderGlobalAware.scoreHolderGlobalIssueDetected( "Multiple 'scoreHolder' global variables found. The global variable is created automatically when a Planning Solution is created. Make sure there are not multiple planning solution objects within the project and that you do not define the 'scoreHolder' global variable manually." );
+                } else if ( scoreHolderFqns.size() == 1 ) {
+                    if ( !supportedScoreHolderTypes.containsAll( scoreHolderFqns ) ) {
+                        scoreHolderGlobalAware.scoreHolderGlobalIssueDetected( "This score level is not supported by current 'scoreHolder' global defined for this project. The global variable is created automatically when a planning solution is created. Select a correct score type in a planning solution or remove this input." );
+                    } else {
+                        scoreHolderGlobalAware.scoreHolderGlobalLoadedCorrectly();
+                    }
+                }
+            }
+        } ).extractProjectScoreTypeFqns( projectPath );
+    }
+
+    public void addPluginToActionList( final Path projectPath,
```

Comment:
This becomes quite "chatty" when opening a Guided Rule Editor instance and you instantiate new instances of the plugins and each is then asked to ```addPluginToActionList(...)``` from drools-wb which in turn does a server-round trip to get the **same** list of **project** "Score Type FQCNs".